llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -1366,6 +1370,52 @@ bool SPIRVInstructionSelector::selectRsqrt(Register ResVReg,
       .constrainAllUses(TII, TRI, RBI);
 }
 
+bool SPIRVInstructionSelector::selectSign(Register ResVReg,
+                                          const SPIRVType *ResType,
+                                          MachineInstr &I) const {
+  assert(I.getNumOperands() == 3);
+  assert(I.getOperand(2).isReg());
+  MachineBasicBlock &BB = *I.getParent();
+  Register InputRegister = I.getOperand(2).getReg();
+  SPIRVType *InputType = GR.getSPIRVTypeForVReg(InputRegister);
+  auto &DL = I.getDebugLoc();
+
+  if (!InputType)
+    report_fatal_error("Input Type could not be determined.");
+
+  bool IsFloatTy = GR.isScalarOrVectorOfType(InputRegister, SPIRV::OpTypeFloat);
+
+  unsigned SignBitWidth = GR.getScalarOrVectorBitWidth(InputType);
+  unsigned ResBitWidth = GR.getScalarOrVectorBitWidth(ResType);
+
+  bool NeedsConversion = IsFloatTy || SignBitWidth != ResBitWidth;
+
+  auto SignOpcode = IsFloatTy ? GL::FSign : GL::SSign;
+  Register SignReg = NeedsConversion
+                         ? MRI->createVirtualRegister(&SPIRV::IDRegClass)
+                         : ResVReg;
+
+  bool Result =
+      BuildMI(BB, I, DL, TII.get(SPIRV::OpExtInst))
+          .addDef(SignReg)
+          .addUse(GR.getSPIRVTypeID(InputType))
+          .addImm(static_cast<uint32_t>(SPIRV::InstructionSet::GLSL_std_450))
+          .addImm(SignOpcode)
+          .addUse(InputRegister)
+          .constrainAllUses(TII, TRI, RBI);
+
+  if (NeedsConversion) {
+    auto ConvertOpcode = IsFloatTy ? SPIRV::OpConvertFToS : SPIRV::OpSConvert;
```

Comment:
This would work if float to int like so:
```hlsl
export int4 fn(float4 p1) {
    return sign(p1);
}
```
But not if the input was double
```hlsl
export int4 fn(double4 p1) {
    return sign(p1);
}
```

https://hlsl.godbolt.org/z/3s5rnfzor

Unless its handled by something else you need a down cast of double to float

Reply:
From the [SPIR-V specification:](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html)
> ... Behavior is undefined if Result Type is not wide enough to hold the converted value. Float Value must be a scalar or vector of floating-point type. It must have the same number of components as Result Type ...

My thinking was that -1.0 / 0.0 / 1.0 should always fit in an `int` and thus we don't have ub.

Reply:
@Keenuts could you take a look  at this PR?  Also let me know what the right behavior should be for this `sign(double)` case?

Reply:
@pow2clk could you weigh in here? I'm not sure what the right thing to do is.

Reply:
I think i'm convinced that this is fine and will sign off.

Reply:
My read of the SPIR-V spec is consistent with @tgymnich's. I am inclined to trust that it can handle directly casting from double to int

Reply:
Note that the latest DXC compiler no longer produces the double->float conversion: https://hlsl.godbolt.org/z/3T5vTv8xb

Reply:
@pow2clk. Awesome I think we can merge this then.

Reply:
Ah here's the hero: https://github.com/microsoft/DirectXShaderCompiler/pull/6533